Dependency Injection

xori opened the pull request:
As recommended in the last [pull request](https://github.com/project-alfox/backend/pull/2#discussion_r195402038) I switched all db accessor factories from static to dependency injected. I'm not totally happy with it, but it gets the job done and I'm open for suggestions to improve it.

In addition the backend now officially has tests.

xori committed:
Dependency Injection

xori commented on src/main/java/com/ace/alfox/controllers/ActionController.java:
```diff
@@ -16,19 +18,28 @@
         ActionFactory doSomething = ActionFactory.named(command).withParameters(params);
+        // Alright, so we've got the IAction in ActionFactory but, because we instantiated it Spring hasn't done any
+        // dependency injection, so we're going to hint that it should do it now.
+        beanFactory.autowireBean(doSomething.action);
```
The "wart" in my opinion.

coltin replied:
Yeah we definitely shouldn't have to do this, especially here in the controller. We can likely find a better way :)

xori replied:
I think once I remove all static functions in `ActionFactory` I'll be able to inject the `beanFactory` right into `ActionFactory` and perform the autowire right [after initialization](https://github.com/project-alfox/backend/pull/3/files/48f077e50210c16f8a80bcfde27eb5bec1134587#diff-0bb83ee27a66a987201591144090b4ecR44)

xori commented on src/main/java/com/ace/alfox/game/MoveAction.java:
```diff
@@ -34,6 +38,11 @@ public ActionResult applyAction(Player player, Map<String, Object> params) {
+        Location location = locationFactory.fetch(player.location);
+        result.log("You see, " + location.title)
+              .log(location.description);
+
```
I wanted to add in a DI data accessor as an example of how it can be used in the Actions

xori commented on src/main/java/com/ace/alfox/lib/ActionFactory.java:
```diff
@@ -1,23 +1,47 @@
+    public static void populateActions(HashMap<String, Class<IAction>> actions) {
+        ActionFactory.actions = actions;
+    }
+
```
This hunk was lifted from `Database` I don't know why it was in there, the act of looking for all the actions should be in the factory that creates it.

xori commented on wercker.yml:
```diff
@@ -1,4 +1,12 @@
+        name: Build and Test
+        code: |
+          echo "Look at me, building and testing..."
+
```
This needs to run the gradle command to test, but I don't know that so that's something to research.

coltin commented on src/main/java/com/ace/alfox/lib/ActionFactory.java:
```diff
@@ -1,23 +1,47 @@
         this.params = p;
     }
 
+    public static void findActions(String namespace) throws ClassNotFoundException {
```
I feel like `findActions` should return something. What do you think of something like `addActionsToFactory` or `addActions`?

coltin commented on src/main/java/com/ace/alfox/lib/ActionFactory.java:
```diff
@@ -1,23 +1,47 @@
+        ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(true);
+        scanner.addIncludeFilter(new AnnotationTypeFilter(PlayerAction.class));
+        for (BeanDefinition bd : scanner.findCandidateComponents(namespace)) {
+            // use my annotation to get simple name
```
Better to say what the annotation is than saying "my annotation".

coltin commented on src/main/java/com/ace/alfox/lib/Database.java:
```diff
@@ -1,48 +1,34 @@
-    public synchronized static DB get() throws IOException {
+    public synchronized DB get(String name) throws IOException {
         if(Database.instance != null)
             return Database.instance;
```
Generally good practice to always use `{}` in Java. Really easy for someone to come along and add an indented print statement or something and break the functionality without realizing it. *horror flashbacks*

xori committed:
DI simplification

xori committed:
instantiate with Spring's BeanFactory

xori committed:
run class discovery at the right time

xori committed:
Update wercker.yml to test project

xori committed:
Update wercker.yml

xori commented:
Alright, @coltin I believe I've addressed all of the commented concerns and am ready to merge.

![image](https://user-images.githubusercontent.com/1091220/41600798-d3ac1c0c-73a4-11e8-90bd-724a3de92dce.png)
I swear that my future pull requests will have less churn. haha

coltin commented on src/main/java/com/ace/alfox/controllers/ActionController.java:
```diff
@@ -3,32 +3,42 @@
+
     @RequestMapping(value = "/perform/{command}", method = { RequestMethod.GET, RequestMethod.POST })
-    public ResponseEntity home(HttpServletRequest request, @PathVariable String command, @RequestBody Map<String, Object> params) throws IOException {
+    public ResponseEntity home(HttpServletRequest request, @PathVariable String command, @RequestBody Map<String, Object> params) {
```
Should rename method from "home" to something like "performCommand"

coltin commented on src/main/java/com/ace/alfox/controllers/ActionController.java:
```diff
@@ -3,32 +3,42 @@
-        Player player = PlayerFactory.fetch(playerId);
-        if(player == null) {
-            return ResponseEntity.notFound().build();
+        NitriteId playerId = (NitriteId) session.getAttribute("pid");
```
"pid" should probably be a constant somewhere. Nbd to leave for now.

coltin commented on src/main/java/com/ace/alfox/controllers/ActionController.java:
```diff
@@ -3,32 +3,42 @@
+        if(playerId == null) {
+            //return ResponseEntity.notFound().build();
+            // Instead of just returning 404, create a test player
+            Player _player = new Player();
```
What does the underscore represent? Is it meant to denote local variables or something?

xori replied:
a habit picked up from javascript, yes a temporary variable only important to this block. In Java I should actually spell it out as such.

coltin commented on src/main/java/com/ace/alfox/controllers/LoginController.java:
```diff
@@ -14,8 +15,7 @@
         HttpSession session = request.getSession(true);
-        session.setAttribute("cid", "1234-1234-1234-1234");
-
+        session.setAttribute("pid", NitriteId.createId(1234123412341234L));
```
Might be worth a comment for why we are currently hardcoding this id

coltin commented on src/main/java/com/ace/alfox/controllers/ActionController.java:
```diff
@@ -3,32 +3,42 @@
 
-        ActionFactory doSomething = ActionFactory.named(command).withParameters(params);
+        Player player = db.players.getById(playerId);
+        ActionFactory doSomething = actionFactory.named(command).withParameters(params);
```
Why is the ActionFactory returning an ActionFactory. Almost seems like it's an ActionFactory Factory?

Seems intuitive to me to rename `withParameters` to `with` and then rename `p` to `params`. So then you end up with:
```java
    public ActionFactory with(Map<String, Object> params) {
        return new ActionFactory(action, params);
    }

    actionFactory.named(command).with(params);
```

Any reason to not combine this?

```java
actionFactory.get(command, params)
```

Where params is optional? I haven't super gone through the code yet so maybe it will be more clear once I do that.

xori replied:
I totally agree. An artifact of the past.

coltin commented on src/main/java/com/ace/alfox/controllers/ActionController.java:
```diff
@@ -3,32 +3,42 @@
+        ActionFactory doSomething = actionFactory.named(command).withParameters(params);
         ActionResult result = player.applyAction(doSomething);
-        PlayerFactory.save(result.player);
+        db.players.update(player);
```
In general I don't think we should expose a classes fields. For example, if we find that `db.players` is suddenly null, it's not easy to track down what code set it to null because anything with a reference to the instance of Database could have accidentally set it to null (or something else). Suddenly the Database class is no longer in control of its fields. Generally I find it's only ever good to expose fields that are immutable. If you hate boilerplate, there are ways around that. For example:

```java
class SomeDataClass {
    @Getter private final String someField;
}
```

This is a Lombok annotation that generates a Getter for this field. You could do the same with `@Setter`. Or you could annotate the whole class with `@Data` and you get getters/setters for all fields with only the one annotation. It will also generate the hashCode, toString, etc for you. I resisted Lombok hard at first, but I've found it makes code very easy to read and so much faster to write, and less buggy. Kind of like using Kotlin I suppose (I've never used Kotlin but I hear similar praises). I find it better than people generating hashCode() with their IDE because if you add a field you don't have to remember to re-generate, which is super easy to miss in a code review.

Anyway, don't feel like you have to change the params to private and start providing getters/setters for everything, it's not a big deal, but I wanted to bring it up. A good compromise if you're against the comments above would be to set these fields to final since I doubt you want things changing them.

coltin replied:
^This might be worth a good debate in person sometime! :)

xori replied:
I agree that all fields exposed should be `final`, a mistake on my part. But I think more than anything I don't dislike the creation of getters, it's the usage of the getters.

A method call to me implies that processing is happening, when direct access to a field is simply navigational.

```java
// the difference between
db.players.getById(1234L);
// and
db.players().getById(1234L);
```

In the latter the getter implies fetching the entire result set, at least to me. I realize that this is just how it is in Java, I'm just used to other languages.

coltin replied:
Interesting! It's so common in Java that I never thought of that relationship. In general I always try to ensure my gets() are cheap, at least on subsequent calls. For example on the dataRepository.getWhatever() it could be an expensive network or database call, but if that's the case I'll cache it so that it's cheap. In this way consumers of my classes always know they are getting as performance gets() as possible. With a get() I can easily decide to make a param lazy, but if my interface is someone accessing the field then I need to populate in the constructor or have them check for initialization and then call some init() function.

I have found in Python and Java it seems super common to just make fields/properties accessible. I'm not even sure if there is a good way to hide them in Python, although maybe you can with underscores or something. I have vague memories of making things private, but maybe that was a dream haha.

coltin commented on src/main/java/com/ace/alfox/game/MoveAction.java:
```diff
@@ -18,7 +20,10 @@
     }};
 
-    public MoveAction() {}
+    Database db;
```
I think this can be `private final`?

coltin commented on src/main/java/com/ace/alfox/game/MoveAction.java:
```diff
@@ -34,6 +39,14 @@ public ActionResult applyAction(Player player, Map<String, Object> params) {
         player.location = player.location.add(directions.get(direction));
         result.log("traveled " + direction);
+
+        Location location = db.locations.find(eq("coordinates", player.location)).firstOrDefault();
```
Would love to see an intermediate layer. `DataRepository` handles all requests for data (whether it's from a cache, database, network, file system, etc). It would have methods like `dataRepository.findLocation(player.location)` and the calling code, `MoveAction` doesn't care about whether it's from Database or anywhere else. Not a blocker for this review, but wanted to get your thoughts on that! :)

I think this would make it really easy to swap out the database (say from LevelDB to Nitrite) without having to change a lot of code, just the DataRepository.

xori replied:
Ha, I was totally thinking this while coding. The Nitrite API is nice, but I totally need to wrap all the repositories with custom fetching logic. The worst bit is that the Nitrate `ObjectRepository` is an interface, so I can't just extend it, I'd need to implement all of the minutia. So it's totally something that needs to happen, I'm just not looking forward to it :P

coltin replied:
For large interfaces I always provide an adapter. ie a class that implements all the methods and does noops, so if you just want to override one method you can do that. Maybe we could add one for ObjectRepository to make it easier to use.

Taking a step back, I don't think you'd have to do anything different to have the DataRepository layer. You would have this code call `dataRepository.findLocation(player.location)` and then datarepo would look like this:

```java
class DataRepository {
    // init stuff

    Location findLocation(Location location) {
       return db.locations.find(eq("coordinates", player.location)).firstOrDefault();
    }
```

I think this is fine. We would need to change all the `eq()` and stuff if we swapped out Nitrite, but it's all in one place and could be replaced slowly. ie you could add a second DB and then replace endpoints as necessary.

coltin commented on src/test/java/com/ace/alfox/MoveActionTests.java:
```diff
@@ -0,0 +1,65 @@
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class MoveActionTests {
```
Tests!!! :D <3

coltin commented on src/test/java/com/ace/alfox/MoveActionTests.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Test
+    public void Move() {
```
Would love to see a test for moving every direction, but not a blocker. This is fine for now!

Capitals on method names seems super weird to me! Though test names I'm used to are `test_whatIsHappening_expectedResult` which is super verbose and everyone I talk to seems to hate it. Though it was great at Yelp; where you viewed test results knew about the format and would print it real nice. Made understanding what broke a lot faster usually.

xori replied:
haha, some C# slipping through. My bad.

Yea I'm fine with that naming scheme, tests are suppose to be verbose and fail loudly. I haven't seen the `_expectedResult` pattern before but can definitely see the appeal. I'm not scared of long function names :P

coltin replied:
Ah yeah I forgot that about C# :D

coltin commented on src/test/java/com/ace/alfox/MoveActionTests.java:
```diff
@@ -0,0 +1,65 @@
+    @Before
+    public void setUp() {
+        player = new Player();
+        move = new MoveAction(db);
```
Do you know if we can autowire this and we just have a test configuration that supplies different instances for tests? For example for prod it provides one DB and for tests it provides an in memory db?

not at all a blocker for this PR, just wanted to ask about it :)

xori replied:
I couldn't figure out a way, to my knowledge the configurations are all file based. We can inject settings via the command line but I don't know how to configure that in the gradle commands. I'm open to suggestions. I dislike _magic strings_.

coltin commented on src/main/java/com/ace/alfox/lib/Database.java:
```diff
@@ -1,56 +1,34 @@
-            System.out.println(alias);
-            Class cl = Class.forName(bd.getBeanClassName());
-            Database.actions.put(alias, cl);
+    public Database(@Value(value = "game") String name) {
```
I think this could use some comments. I wasn't really sure what was going on until I saw the tests and then I was like AHA.

xori replied:
Fair

xori committed:
Requested adjustments.

coltin reviewed (approved):
Ship it! Sorry this lingered for so long, I'll prioritize code reviews in the future
